src/shogun/mathematics/linalg/GPUMemoryViennaCL.h

```diff
@@ -0,0 +1,88 @@
+#ifndef GPUMemory_ViennaCL_H__
+#define GPUMemory_ViennaCL_H__
+
+#ifdef HAVE_CXX11
+#ifdef HAVE_VIENNACL
+
+#include <viennacl/vector.hpp>
+#include <memory>
+
+namespace shogun
+{
+
+template <typename T>
+struct GPUMemoryViennaCL : public GPUMemoryBase<T>
+{
+	typedef viennacl::backend::mem_handle VCLMemoryArray;
+	typedef viennacl::vector_base<T, std::size_t, std::ptrdiff_t> VCLVectorBase;
+
+	GPUMemoryViennaCL()
+	{
+		init();
+	};
+
+	GPUMemoryViennaCL(GPUMemoryBase<T>* vector)
```

Comment:
I would only accept GPUMemoryViennaCL here and force the cast to be done by caller